File: backend/msreserva/src/main/java/com/dac/msreserva/listeners/TransactionListener.java
package com.dac.msreserva.listeners;

import org.springframework.amqp.rabbit.annotation.RabbitListener;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;
import com.dac.msreserva.DTO.ReservaTransactionDTO;
import com.dac.msreserva.DTO.VooDTO;
import com.dac.msreserva.services.ReservaService;
import com.fasterxml.jackson.databind.ObjectMapper;

@Service
public class TransactionListener {

    private final ReservaService service;
    @Autowired
    private ObjectMapper objectMapper;

    public TransactionListener(ReservaService service) {
        this.service = service;
    }

    @RabbitListener(queues = "criareserva.reserva")
    public String efetuarReserva(String payload) {
        try {
            System.out.println("Criar reservada escutado");
            System.out.println(payload);
            ReservaTransactionDTO reservaTransaction = objectMapper.readValue(payload, ReservaTransactionDTO.class);

            Object resultado = service.efetuarReserva(reservaTransaction);
            System.out.println("Retorno");
            System.out.println(resultado);
            return objectMapper.writeValueAsString(resultado);
        } catch (Exception e) {
            throw new RuntimeException("Erro ao processar reserva", e);
        }
    }

    @RabbitListener(queues = "cancelareserva.reserva")
    public String cancelarReserva(String payload) {
        try {
            System.out.println("Iniciando cancelamento da reserva");

            String codigo = objectMapper.readValue(payload, String.class);
            System.out.println("Código recebido: " + codigo);

            Object reserva = service.cancelarReserva(codigo);
            return objectMapper.writeValueAsString(reserva);

        } catch (ResponseStatusException e) {
            return "{\"erro\": \"" + e.getReason() + "\", \"status\": " + e.getStatusCode().value() + "}";
        } catch (Exception e) {
            e.printStackTrace();
            return "{\"erro\": \"Erro inesperado ao cancelar reserva\"}";
        }
    }

    @RabbitListener(queues = "realizavoo.reserva")
    public void realizaVoo(String payload) {
        try {
            System.out.println("Realizar voo escutado");
            VooDTO voo = objectMapper.readValue(payload, VooDTO.class);
            service.realizaVoo(voo);

        } catch (Exception e) {
            System.err.println("Erro ao processar realização de voo: " + e.getMessage());
            e.printStackTrace();
        }
    }

    @RabbitListener(queues = "cancelavoo.reserva")
    public void canceladoVoo(String payload) {
        try {
            System.out.println(("Cancelar voo escutado"));
            VooDTO voo = objectMapper.readValue(payload, VooDTO.class);
            service.canceladoVoo(voo);
        } catch (Exception e) {
            System.err.println("Erro ao processar cancelamento de voo: " + e.getMessage());
            e.printStackTrace();
        }
    }
}

File: backend/reserva-service/src/main/kotlin/br/ufpr/dac/reserva_service/resource/TransactionListener.kt
package br.ufpr.dac.reserva_service.resource

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import org.springframework.amqp.rabbit.annotation.RabbitListener
import org.springframework.stereotype.Service
import utils.dto.RabbitMessageDTO
import utils.dto.ReservaTransactionDTO
import utils.dto.VooOutputDTO
import utils.gson.ZonedDateTimeAdapter
import java.time.ZonedDateTime

@Service
class TransactionListener(private val service: ReservaService) {
    private final val gson: Gson = GsonBuilder()
        .registerTypeAdapter(ZonedDateTime::class.java, ZonedDateTimeAdapter())
        .create()

    @RabbitListener(queues = ["emiratads.criareserva.reserva"], errorHandler = "customErrorHandler")
    fun efetuarReserva(payload: String): String {
        val reservaTransaction = gson.fromJson(payload, ReservaTransactionDTO::class.java)

        val milhasTransaction = service.efetuarReserva(reservaTransaction)
        val response = RabbitMessageDTO(true, milhasTransaction)
        return gson.toJson(response)
    }

    @RabbitListener(queues = ["emiratads.cancelareserva.reserva"], errorHandler = "customErrorHandler")
    fun cancelarReserva(codigo: String): String {
        val reserva = service.cancelarReserva(codigo)
        val response = RabbitMessageDTO(true, reserva)

        return gson.toJson(response)
    }

    @RabbitListener(queues = ["emiratads.cancelavoo.reserva"], errorHandler = "customErrorHandler")
    fun canceladoVoo(payload: String): String {
        val voo = gson.fromJson(payload, VooOutputDTO::class.java)
        val reserva = service.canceladoVoo(voo)
        val response = RabbitMessageDTO(true, reserva)

        return gson.toJson(response)
    }

    @RabbitListener(queues = ["emiratads.realizavoo.reserva"], errorHandler = "customErrorHandler")
    fun realizaVoo(payload: String): String {
        val voo = gson.fromJson(payload, VooOutputDTO::class.java)
        val reserva = service.realizaVoo(voo)
        val response = RabbitMessageDTO(true, reserva)

        return gson.toJson(response)
    }
}